File: javascript/json-transform/src/functions/TransformerFunctionMerge.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import { JsonMergeOptions, merge } from "../JsonHelpers";
import { AsyncSequence, asyncSequenceOf } from "@wortise/sequency";

class TransformerFunctionMerge extends TransformerFunction {
  constructor() {
    super({
      argsSets: [
        [
          { name: "deep", type: ArgType.Boolean, defaultValue: false },
          { name: "arrays", type: ArgType.Boolean, defaultValue: false },
        ],
      ],
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const streamer = await context.getJsonElementStreamer(null);
    if (streamer == null || streamer.knownAsEmpty()) return null;

    const deep = await context.getBoolean("deep");
    const arrays = await context.getBoolean("arrays");

    const options: JsonMergeOptions = {
      deep: deep ?? undefined,
      concatArrays: arrays ?? undefined,
    };

    let result: Record<string, any> = {};

    return (asyncSequenceOf(asyncSequenceOf(result), streamer.stream()).flatten() as AsyncSequence<any>).reduce(
      async (acc, value) => {
        return merge(result, value, options);
      },
    );
  }
}

export default TransformerFunctionMerge;

File: java/json-transform/src/main/java/ai/bizone/jsontransform/functions/TransformerFunctionMerge.java
package ai.bizone.jsontransform.functions;

import ai.bizone.jsontransform.adapters.JsonAdapter;
import ai.bizone.jsontransform.functions.common.*;
import co.nlighten.jsontransform.functions.common.*;

import java.util.Map;
import java.util.stream.Stream;

public class TransformerFunctionMerge extends TransformerFunction {

    public TransformerFunctionMerge() {
        super(FunctionDescription.of(
            Map.of(
                "deep", ArgumentType.of(ArgType.Boolean).position(0).defaultValue(false),
                "arrays", ArgumentType.of(ArgType.Boolean).position(1).defaultValue(false)
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        var streamer = context.getJsonElementStreamer(null);
        if (streamer == null || streamer.knownAsEmpty())
            return null;

        var deep = context.getBoolean("deep");
        var arrays = context.getBoolean("arrays");

        var options = new JsonAdapter.JsonMergeOptions(deep, arrays);

        var adapter = context.getAdapter();
        var result = adapter.createObject();

        return ((Stream<Object>)streamer.stream())
                .reduce(result, (acc, value) -> adapter.merge(acc, value, options));
    }
}
